src/lib/importSystem/record.ts

```diff
@@ -0,0 +1,562 @@
+ */
+const importHasMoreToProcess = async (
+  importId: number,
+) => {
+  const pendingFiles = await ImportFile.findAll({
+    attributes: ['id'],
+    where: {
+      importId,
+      processAttempts: { [Op.lt]: 5 },
+      status: [IMPORT_STATUSES.COLLECTED, IMPORT_STATUSES.PROCESSING_FAILED],
+    },
+  });
+
+  return pendingFiles?.length || 0 >= 1;
+};
+
+/**
+ * Retrieves the next import file to process based on the provided import ID and maximum attempts.
+ *
+ * @param importId - The ID of the import.
+ * @param maxAttempts - The maximum number of process attempts allowed. Defaults to 5 if
+ * not provided.
+ * @returns A promise that resolves to the next import file to process, or undefined if no
+ * import file is found.
+ */
+const getNextFileToProcess = async (
+  importId: number,
+  maxAttempts = 5,
+) => {
+  // Find the next import file to process
+  const importFile = await ImportFile.findOne({
+    attributes: [
+      ['id', 'importFileId'],
+      'fileId',
+      'status',
+      'processAttempts',
+    ],
+    include: [
+      {
+        model: File,
+        as: 'file',
+        attributes: [
+          'key',
+        ],
+      },
+      {
+        model: Import,
+        as: 'import',
+        attributes: [
+          'definitions',
+        ],
+      },
+    ],
+    where: {
+      importId,
+      [Op.or]: [
+        { status: IMPORT_STATUSES.COLLECTED }, // Import file is in the "collected" status
+        {
+          // Import file is in the "processing_failed" status
+          status: IMPORT_STATUSES.PROCESSING_FAILED,
+          // Number of process attempts is less than the maximum allowed attempts
+          processAttempts: { [Op.lt]: maxAttempts },
+        },
+      ],
+    },
+    order: [
+      ['ftpFileInfo->>"name"', 'ASC'], // Order by file name in ascending order
+      ['ftpFileInfo->>"date"', 'ASC'], // Order by file date in ascending order
+    ],
+    limit: 1, // Limit the result to 1 record
+  });
+
+  return importFile;
+};
+
+/**
+ * Records available files for a specific import.
+ * @param importId - The ID of the import.
+ * @param availableFiles - An array of available files with their details.
+ * @returns A promise that resolves when all database operations are completed.
+ */
+const recordAvailableFiles = async (
+  importId: number,
+  availableFiles: {
+    fullPath: string,
+    fileInfo: FTPFileInfo,
+    stream?: Promise<Readable>,
+  }[],
+) => {
+  // Retrieve current import files from the database
+  const currentImportFiles: {
+    id: number,
+    importId: number,
+    fileInfo: FTPFileInfo,
+  }[] = await ImportFile.findAll({
+    attributes: [
+      'id',
+      'importId',
+      ['ftpFileInfo', 'fileInfo'],
+    ],
+    where: {
+      importId,
+      status: { [Op.not]: IMPORT_STATUSES.PROCESSED },
+    },
+    raw: true,
+  });
+
+  // Separate the available files into new, matched, and removed files
```

Comment:
I don't understand this construction - why are you initializing these inside of an array and then immediately destructuring? 

You could also extract the ```some```, that might clean it up a little, it's hard to parse now.

```js
//name is just a suggestion
const fileMatchesImportIdAndHasPathAndName = (currentImportFile) => (
          importId === currentImportFile.importId
          && availableFile.fileInfo.path === currentImportFile.fileInfo.path
          && availableFile.fileInfo.name === currentImportFile.fileInfo.name
        );


// New files are those that are not already recorded in the database
const newFiles = availableFiles
      .filter((availableFile) => !currentImportFiles
        .some(fileMatchesImportIdAndHasPathAndName);
        
// Matched files are those that are already recorded in the database
const matchedFiles = availableFiles
      .filter((availableFile) => currentImportFiles
        .some(fileMatchesImportIdAndHasPathAndName);
        
        
//        ...etc

Reply:
It was a pattern I got into while doing the GR to keep steps aligned. But in this case I can see it does not add much  value.